### DISCUSSION TITLE
Are Queue operations protected against simultaneous access?

### DISCUSSION THREAD

**Discussion by gmag11** [Mair/esp32-course#117] [Course Qeustions] [answered]
I'm testing some code with a data generator a consumer with a queue in between. My intention is to measure max performance in my code.

If I write into the queue slowly (every some tenths milliseconds) everything goes ok. But if elements are written faster I get an assert error quite at the beginning.

I guess that 2nd queue send happens meanwhile queue reeceive has not finished. Should I protect send and receive for mutual exclusion? Is it managed inside queue code?

**Comment by Mair** [comment, marked as answer]
@gmag11 No internal protection that I know of. If you have 2 tasks sending data to a receiver at the same time, It's Probably a good idea to wrap your receiver in a mutex

**Comment by gmag11** [comment]
Hi, I've been looking at FreeRTOS source code. It seems that send is protected under a critical section: https://github.com/maniacbug/FreeRTOS/blob/master/queue.c#L895
This would mean that no mutex protection is needed in addition, wouldn't it?

**Reply by Mair** [reply]
@gmag11 You could be right on this. In which case your error might be caused by something else. sorry for not being more helpful

**Reply by gmag11** [reply]
Error was caused by a null pointer. Nothing to do with mutual access.
